### PR TITLE
Feat/ota 2159/delegations

### PR DESCRIPTION
**Pull request by patrickvacek**
I've finally finished all the tests I could think of for basic delegations:

1. DelegationSimpleUptane (successful)
1. DelegationRedundantUptane (successful; target is listed in top-level targets and delegation)
1. DelegationPathMismatchUptane
1. DelegationKeyMissingUptane
1. DelegationUnsignedUptane
1. DelegationBadKeyIdsUptane
1. DelegationMissingUptane
1. DelegationEmptyUptane
1. DelegationHashMismatchUptane
1. DelegationExpiredUptane
1. DelegationSnapshotMissingUptane (delegation not listed in snapshot)
1. DelegationSnapshotVersionMismatchUptane
1. DelegationZeroThresholdUptane
1. DelegationUnmetThresholdUptane
1. DelegationTwoSignaturesUptane (successful)
1. DelegationNonUniqueSignaturesUptane

As well as one thing that wasn't tested before but I added while doing the above: ImageRepoSnapshotTargetsVersionMismatchUptane.

I did not add tests for bad hashes or lengths in the snapshot because the former shouldn't be there (see [here](https://github.com/uptane/uptane-standard/issues/90) and [here](https://github.com/uptane/uptane-standard/pull/92)) and the former probably isn't necessary either. If we decide to keep it, it'll be easy to add a test for it.

I also did not add tests for nested delegations (because I haven't gotten around to it and the backend doesn't support them yet), delegation revocation (same reasons), or multi-role delegations (because they aren't implemented anywhere yet).

Thanks to @heartsucker for getting the ball rolling on starting this almost a year ago!

### CODE REVIEW

**Comment by patrickvacek** [comment]
See also https://github.com/advancedtelematic/aktualizr/pull/1157.

(Also tagging @tkfu to get someone with some deep Uptane insight to take a quick look.)

File: tuf_vectors/uptane.py [hunk 1]
```diff
@@ -2702,3 +2702,1065 @@ def __targets(hardware_id: str, ecu_identifier: str=None) -> list:
...
+        }
+
+    STEPS = [
+        (DirectorStep, ImageStep),
+    ]
+
+
+class DelegationExpiredUptane(Uptane):
+
+    '''The delegation metadata has expired'''
```
Comment (tkfu) [inline tuf_vectors/uptane.py]: I'd call this "delegated metadata". To me, "delegation metadata" is the thing that points you to delegated metadata.

File: tuf_vectors/uptane.py [hunk 2]
```diff
@@ -2702,3 +2702,1065 @@ def __targets(hardware_id: str, ecu_identifier: str=None) -> list:
...
+        }
+
+    STEPS = [
+        (DirectorStep, ImageStep),
+    ]
+
+
+class DelegationNonUniqueSignaturesUptane(Uptane):
+
+    '''A delegation has duplicate signatures'''
```
Comment (tkfu) [inline tuf_vectors/uptane.py]: Maybe something like "the threshold of signatures is met, but the signatures are non-unique"

**Comment by patrickvacek** [comment]
Fixed the first issue. Left a comment about the second one (non-unique signatures) because it requires further consideration and is out of scope for this. I made a ticket FWIW: https://saeljira.it.here.com/browse/OTA-2448.

**Pull request merged**